plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/gitlab.ts

```diff
@@ -38,14 +38,33 @@ export function createPublishGitlabAction(options: {
   return createTemplateAction<{
     repoUrl: string;
     defaultBranch?: string;
+    /** @deprecated in favour of settings.visibility field */
     repoVisibility?: 'private' | 'internal' | 'public';
```

Comment:
I feel that we lose a lot of type information by accepting `Record<string, any>`. I suggest we retain it in the new field, same thing goes for topics.

Reply:
Yes, you are correct. Gitlab API supports a lot of parameters we can provide during project creation - https://docs.gitlab.com/ee/api/projects.html#create-project. Ideally, action should support all of them for better flexibility and extensibility. Here I had two options:

1. provide all currently supported parameters
2. just use a hashmap

Each of the options has pros and cons:

Option 1:
+ type safety
+ Convenient to use

- too many parameters
- some of the parameters are specific to versions and distributions of the gitlab, so it will be hard to support
- in case some parameters are added or deleted we will need to adjust the action

Option 2:
+ easy configuration
+ compatibility with any version of gitlab
+ no need to make changes when gitlab API evolves

- no type safety
- anyone can pass any data to gitlab API

what are your thoughts?

Reply:
We should be able to do something in between by also adding visibility to the Record type.
Basically something like this which would give us type safety and also the flexibility to use unknown flags. We try really hard to stay away from `any` types in the API as it get's very hard to evolve and maintain. I would argue that it might be better to add the new settings when needed and not open up the API too much as this API is 1.0 and considered stable. @benjdlambert do you have any opinions about this one?

```ts
settings: Record<string, any> & {
    repoVisibility?: 'private' | 'internal' | 'public';
  };
```

Reply:
@benjdlambert sorry for pinging you, we need your help here

Reply:
Hmm - yeah, this is kind of unfortunate not to have typesafety here, but I understand that theres a lot for us to manually type if we want to do that. It would be awesome if we could extract the types from the gitbeaker API, or get a jsonschema for the `gitlab` API for this. 

Feeling that a better type if we were to use this would be `JSONObject` or `Record<string, JSONValue>` rather than `any`. 

I'm still feeling though that it's better for us to type these somehow, because whilst it creates an API contract that we have control over. If the gitbeaker library changes underneath, then at least we don't have to get templates authors to update all their templates to change a setting name when we can do the mapping ourselves maybe? 

What do you think @ikatlinsky?

Reply:
@benjdlambert @jhaals thanks for your replies, I agree with your desire, what I will do

1. i will try to generate a typesafe contract from gitbreaker api
2. i will use results to make action API typesafe
3. i will not make a direct link to the gitbreaker or gitlab APIs, will use it as a starting point only
4. in case gitbreaker or gitlab API will change - mapping for parameters will be updated inside action to support backward compatibility for action users

in case 1 option will be hard to achieve i will type it manually. Sounds good?

Reply:
sounds good :tada:

Reply:
looked at how gitbreaker api and actually it does not type anything, except for name and path parameters and they actually expect just Record<string, unknown>, will try to add typing manually.